pkg/cmd/gen_driftignore.go

```diff
@@ -16,19 +17,34 @@ func NewGenDriftIgnoreCmd() *cobra.Command {
 	cmd := &cobra.Command{
 		Use:   "gen-driftignore",
 		Short: "Generate a .driftignore file based on your scan result",
-		Long:  "This command will generate a new .driftignore file containing your current drifts and send output to /dev/stdout\n\nExample: driftctl scan -o json://stdout | driftctl gen-driftignore -i /dev/stdin > .driftignore",
+		Long:  "This command will generate a new .driftignore file containing your current drifts\n\nExample: driftctl scan -o json://stdout | driftctl gen-driftignore",
 		Args:  cobra.NoArgs,
 		RunE: func(cmd *cobra.Command, args []string) error {
-			if opts.InputPath == "" {
-				return errors.New("Error: you must specify an input to parse JSON from. Use driftctl gen-driftignore -i <drifts.json>\nGenerate a JSON file using the output flag: driftctl scan -o json://path/to/drifts.json")
+			driftFile := os.Stdin
+			if opts.InputPath != "-" {
+				var err error
+				driftFile, err = os.Open(opts.InputPath)
+				if err != nil {
+					return err
+				}
+				defer driftFile.Close()
 			}
 
-			_, list, err := genDriftIgnore(opts)
+			_, list, err := genDriftIgnore(opts, driftFile)
 			if err != nil {
 				return err
 			}
 
-			fmt.Println(list)
+			ignoreFile := os.Stdout
+			if opts.OutputPath != "-" {
+				var err error
+				ignoreFile, err = os.OpenFile(opts.OutputPath, os.O_CREATE|os.O_TRUNC|os.O_WRONLY, 0644)
```

Comment:
Perhaps it's better to _not_ truncate and overwrite existing files, and use `O_EXCL` instead of `O_TRUNC`

Reply:
Yeah now that we talk about it, I think EXCL is the safest. Make the user explicitly remove the file if they want to overwrite it. I'm not sure why we'd use APPEND?

Reply:
Ah wait, I think I might see why APPEND could work: if you already have a `.driftignore`, the new drift won't contain the already-ignored resources... is that what you meant?

This is more complicated than I'd thought 😁

Reply:
> I'm not sure why we'd use APPEND?

You could run it again to update your .driftignore without re running a full scan. From what I know @sjourdan used to >> into the file (and filled a bug because there was no \n at the end of the generated file). But yeah now we talk about this I'm not sure what use case it would fit so exclude would be safe.

Reply:
That makes sense... I'll defer to your expertise here, I've barely used driftctl yet. So far my thoughts are:

* append would definitely satisfy some use-cases, but I haven't fully considered whether it'd be a footgun for other use-cases.
* excl would force the user to perform a full rescan with no ignore rules (unless they pass `--driftignore` to work around this), which sort of feels like the most conservative approach.
* truncate is probably wrong... we'd be implicitly removing the existing driftignore rules!

Reply:
I think truncate should be dropped unless we introduce a previous file backup mechanism. If we trunc then we should spawn a `.driftignore.bak` file.

Anyway I think the debate here is to decide if we want to keep the append behavior or not, with the UX implication @craigfurman greatly summarized above ☝🏻 🙏🏻 

>  - append would definitely satisfy some use-cases, but I haven't fully considered whether it'd be a footgun for other use-cases.
>  - excl would force the user to perform a full rescan with no ignore rules (unless they pass --driftignore to work around this), which sort of feels like the most conservative approach.
we could even support the append case by adding an extra optional flag --append

I think the append way is more convenient and more user friendly, let me elaborate.

driftctl is most of the time used in an iterative way. You run, check your result, fix / ignore, and repeat.
I think it fit better with this usage, I mean the UX will be better since you do not need to invalidate your previous ignore policy. In term of performance this is probably also a better move since driftignore rules are taken into account when deciding to make API calls or not (this is not the case with `--filter`). It work fine for a first run too.

I'm seeing only one drawback of this approach. For example if we take the case when you decided to have an inverted ignore file logic, like this:

```
*
!aws_s3_bucket
```

You may see unmanaged bucket in your output after a scan, pipe this to the gen-driftignore cmd and you'll end up with this kind of file:

```
*
!aws_s3_bucket
aws_s3_bucket.my-unmanaged-bucket
```

I don't know how we handle this but I think that the `!aws_s3_bucket` rule is taking the precedence over this one `aws_s3_bucket.my-unmanaged-bucket`. So we need people to reorder their logic, but I'm not seeing a simple way to handle that.

Reply:
Thanks for that @eliecharra, I think I understand now, and agree with you. WDYT of using append, _and_ backing up the output file (`.driftignore` by default) to `${NAME}.bak` (unless the file is stdin), just to be on the safe side?

Even if we default to append, there are a couple of potential rough edges remaining, including the one you call out at the end of your last comment. At the moment, section comments like `# Resources not covered by IaC` are duplicated when we append, but this is also the case when using shell `>>`, or manually merging files. I reckon changing things like that are probably out of scope of this particular PR, WDYT?

Finally, if we're writing to a file, we could print a message to stderr informing the user that we did this, in case they get confused by the silence. I'm not sure whether or not I like that idea in general, but since we're proposing to change the default behaviour, it might make sense. Of course totally up to you.

Reply:
> WDYT of using append, and backing up the output file (.driftignore by default) to ${NAME}.bak (unless the file is stdin), just to be on the safe side?

For me this is as minor safety feature since there is no real risk to blow existing rules. Maybe you can open an issue for it (it could be a good first issue since it is a simple nice to have feat) 😃 

> Even if we default to append, there are a couple of potential rough edges remaining, including the one you call out at the end of your last comment. At the moment, section comments like # Resources not covered by IaC are duplicated when we append, but this is also the case when using shell >>, or manually merging files. I reckon changing things like that are probably out of scope of this particular PR, WDYT?

Maybe we could add a header like `# Generated by gen-driftignore cmd @05/11/2021 10:12` while appending results ? So we could have multiples sequential ignore blocks with `# Resources not covered by IaC` acting as sub-section. In a iterative process it could be useful for used to be able to get the history WDYT ?

> Finally, if we're writing to a file, we could print a message to stderr informing the user that we did this, in case they get confused by the silence. I'm not sure whether or not I like that idea in general, but since we're proposing to change the default behaviour, it might make sense. Of course totally up to you.

If we are writing to a file, why not using stdout instead of stderr ?

Reply:
> For me this is as minor safety feature since there is no real risk to blow existing rules. Maybe you can open an issue for it

Sure, so just to be clear you want to leave that out of this PR? I'm fine with that 🙂 

> Maybe we could add a header like # Generated by gen-driftignore cmd @05/11/2021 10:12

That's a great idea, and as you say a good first step towards maybe doing some auto-merging in the future, while not confusing users today.

> If we are writing to a file, why not using stdout instead of stderr ?

Sure, if we're writing a file then we could get away with printing diagnostic messages to stdout, but traditionally that's what stderr is for. I think it confers a minor advantage: if a user _does_ try to pipe the output into something that parses it, or redirects it to a file, e.g. because they're expecting the old behaviour of driftctl, they'll still see that diagnostic message on stderr (as long as they don't pipe that too). A minor point for sure, but I don't think we _lose_ anything by using stderr so it makes sense to me to take the very small win! WDYT?

Reply:
> Sure, so just to be clear you want to leave that out of this PR? I'm fine with that slightly_smiling_face

Yeah ! We tried to leave these kind of stuff to the community. We wanted to encourage contributions and BTW it's a win-win for everyone 😃 

> That's a great idea, and as you say a good first step towards maybe doing some auto-merging in the future, while not confusing users today.

Yeah exactly auto merging could be the ultimate level of this command, we'll be able to handle inverted logic by prepending some rules instead of appending. This one will probably need a `.bak` file BTW 😛 

> Sure, if we're writing a file then we could get away with printing diagnostic messages to stdout, but traditionally that's what stderr is for. I think it confers a minor advantage: if a user does try to pipe the output into something that parses it, or redirects it to a file, e.g. because they're expecting the old behaviour of driftctl, they'll still see that diagnostic message on stderr (as long as they don't pipe that too). A minor point for sure, but I don't think we lose anything by using stderr so it makes sense to me to take the very small win! WDYT?

I think you have a better experience than me on that so I leave that up to you, I'm fine to use both. 

More generally about BC break: for now the project is in a really early beta stage
We do not have shipped a `1.x.x` so we can totally assume BC break in minor versions. Anyway there is no reason to introduce some if we can avoid OFC 😃